README.md
# tempera

[![Release](https://img.shields.io/github/release/ShogunPanda/tempera.svg)](https://github.com/ShogunPanda/tempera/releases/latest)
[![GoDoc](https://godoc.org/github.com/ShogunPanda/tempera?status.svg)](https://godoc.org/github.com/ShogunPanda/tempera)
[![Go Report Card](https://goreportcard.com/badge/github.com/ShogunPanda/tempera)](https://goreportcard.com/report/github.com/ShogunPanda/tempera)
[![Build Status](https://img.shields.io/travis/rust-lang/rust/master.svg)](https://travis-ci.org/ShogunPanda/tempera)
[![License](https://img.shields.io/github/license/ShogunPanda/tempera.svg)](https://github.com/ShogunPanda/tempera/blob/master/LICENSE.md)

Template based terminal coloring made really easy.

http://sw.cowtech.it/tempera

## Usage

Tempera allows to add coloring to terminal in a really easy way.

### Colorize strings

To colorize strings, simply use the [Colorize](https://godoc.org/github.com/ShogunPanda/tempera#Colorize) function, passing a list of styles you want to apply.
The list of supported color names correspondes to the keys of [ANSICodes](https://godoc.org/github.com/ShogunPanda/tempera#pkg-variables) variable.
```go

import "github.com/ShogunPanda/tempera"

inRed := tempera.Colorize("Colorized", "red")
inRedWithBlueBackground := tempera.Colorize("Colorized", "red bgBlue")
```

### Colorize templates

To colorize a template using a tagged template syntax, simply use the [ColorizeTemplate](https://godoc.org/github.com/ShogunPanda/tempera#ColorizeTemplate) function.

```go

import "github.com/ShogunPanda/tempera"

colored := tempera.ColorizeTemplate("{red}This is in red, {green underline}this in green underlined{-}, this in red again.")
```

The template recognizes styles between curly braces (use a double opening brace to escape them) and the token `{-}` as universal closing tag (which also restores the previous style).

The closing tag at the end of the string can be omitted, since tempera will append the global reset style (`\x1b[0m`) if any style was set.

If you want to discard styles to be restored, use the `{reset}` token.

### Setting custom styles

If you want to define custom styles, use the [AddCustomStyle](https://godoc.org/github.com/ShogunPanda/tempera#AddCustomStyle) function.

```go
import "github.com/ShogunPanda/tempera"

tempera.AddCustomStyle("important", "red underline")
colored := tempera.ColorizeTemplate("This is in red, underlined.", "important")
```

### 256 and 16 millions colors support

tempera supports 256 ANSI codes and 16m RGB colors. Just give it a try:

```go
import "github.com/ShogunPanda/tempera"

fgAnsi256 := tempera.Colorize("color me", "ansi:100")
bgAnsi256 := tempera.Colorize("color me", "bgANSI:3,4,5")

fgRgb := tempera.Colorize("color me", "rgb:255,0,0")
bgRgb := tempera.Colorize("color me", "bgRGB:0,255,0")

fgHex := tempera.Colorize("color me", "hex:#FF0000")
bgHex := tempera.Colorize("color me", "bgHEX:00FF00")
```

ANSI, RGB, and HEX can be used in style definitions and templates as well.

## API Documentation

The API documentation can be found [here](https://godoc.org/github.com/ShogunPanda/tempera).

## Contributing to tempera

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.

## Copyright

Copyright (C) 2018 and above Shogun (shogun@cowtech.it).

Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.



CHANGELOG.md
### 2018-02-04 / 1.1.0

* Added `CleanTemplate` function.

### 2018-02-04 / 1.0.0

* Initial version.

main_test.go
/*
 * This file is part of tempera. Copyright (C) 2018 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.
 */

package tempera

import (
	"errors"
	"testing"

	. "github.com/franela/goblin"
)

func Test(t *testing.T) {
	g := Goblin(t)

	g.Describe("Colorize", func() {
		g.It("Applies known styles", func() {
			g.Assert(Colorize("ABC", "bgBlack", "red")).Equal("\x1b[40m\x1b[31mABC\x1b[39m\x1b[49m")
		})

		g.It("Ignores known styles", func() {
			g.Assert(Colorize("ABC", "whatever", "red")).Equal("\x1b[31mABC\x1b[39m")
		})

		g.It("Supports ANSI 256 colors, ignoring invalid colors", func() {
			g.Assert(Colorize("ABC", "ANSI:232")).Equal("\x1b[38;5;232mABC\x1b[39m")
			g.Assert(Colorize("ABC", "bgANSI:333")).Equal("ABC")
			g.Assert(Colorize("ABC", "bgansi:2,4,0")).Equal("\x1b[48;5;112mABC\x1b[49m")
			g.Assert(Colorize("ABC", "ANSI:2,4,6")).Equal("ABC")
		})

		g.It("Supports ANSI 16m RGB colors, ignoring invalid colors", func() {
			g.Assert(Colorize("ABC", "rgb:255,232,0")).Equal("\x1b[38;2;255;232;0mABC\x1b[39m")
			g.Assert(Colorize("ABC", "bgRGB:33,66,99")).Equal("\x1b[48;2;33;66;99mABC\x1b[49m")
			g.Assert(Colorize("ABC", "bgRGB:33,123,3000")).Equal("ABC")
		})

		g.It("Supports ANSI 16m HEX colors, ignoring invalid colors", func() {
			g.Assert(Colorize("ABC", "hex:F0d030")).Equal("\x1b[38;2;240;208;48mABC\x1b[39m")
			g.Assert(Colorize("ABC", "bgHEX:0099FF")).Equal("\x1b[48;2;0;153;255mABC\x1b[49m")
			g.Assert(Colorize("ABC", "bgHEX:0099GG")).Equal("ABC")
		})
	})

	g.Describe("ColorizeTemplate", func() {
		g.It("Applies known styles and closes them in the right order", func() {
			g.Assert(ColorizeTemplate("{red}ABC{green}CDE{-}EFG{-}HIJ")).Equal("\x1b[31mABC\x1b[32mCDE\x1b[39m\x1b[31mEFG\x1b[39mHIJ\x1b[0m")
		})

		g.It("Ignores unknown styles", func() {
			g.Assert(ColorizeTemplate("{red}ABC{yolla}CDE{-}EFG{-}HIJ")).Equal("\x1b[31mABCCDE\x1b[39mEFGHIJ\x1b[0m")
		})

		g.It("Ignores unbalanced parenthesis", func() {
			g.Assert(ColorizeTemplate("{red}}ABC{-}")).Equal("\x1b[31m}ABC\x1b[39m\x1b[0m")
		})

		g.It("Ignores unbalanced tags", func() {
			g.Assert(ColorizeTemplate("{red}ABC")).Equal("\x1b[31mABC\x1b[0m")
		})

		g.It("Double curly braces are respected", func() {
			g.Assert(ColorizeTemplate("{{red}")).Equal("{red}")
		})

		g.It("Closing tag ignores further specs", func() {
			g.Assert(ColorizeTemplate("{red}ABC{green}CDE{- yellow}EFG{-}HIJ")).Equal("\x1b[31mABC\x1b[32mCDE\x1b[39m\x1b[31mEFG\x1b[39mHIJ\x1b[0m")
		})

		g.It("Reset tag cleans the stack", func() {
			g.Assert(ColorizeTemplate("{red}ABC{green}CDE{reset red}EFG{-}HIJ")).Equal("\x1b[31mABC\x1b[32mCDEEFGHIJ\x1b[0m")
		})

		g.It("Supports ANSI, RGB and HEX colors", func() {
			g.Assert(ColorizeTemplate("{ANSI:5,0,0}ABC{RGB:0,255,0}CDE{bgHEX:#0000FF}EFG")).Equal("\x1b[38;5;196mABC\x1b[38;2;0;255;0mCDE\x1b[48;2;0;0;255mEFG\x1b[0m")
		})
	})

	g.Describe("CleanTemplate", func() {
		g.It("Removes style tags from a template", func() {
			g.Assert(CleanTemplate("{red}ABC{green}CDE{-}EFG{-}HIJ")).Equal("ABCCDEEFGHIJ")
			g.Assert(CleanTemplate("{red}ABC{yolla}CDE{-}EFG{-}HIJ")).Equal("ABCCDEEFGHIJ")
			g.Assert(CleanTemplate("{red}}ABC{-}")).Equal("}ABC")
			g.Assert(CleanTemplate("{red}ABC")).Equal("ABC")
			g.Assert(CleanTemplate("{{red}")).Equal("{red}")
			g.Assert(CleanTemplate("{red}ABC{green}CDE{- yellow}EFG{-}HIJ")).Equal("ABCCDEEFGHIJ")
			g.Assert(CleanTemplate("{red}ABC{green}CDE{reset red}EFG{-}HIJ")).Equal("ABCCDEEFGHIJ")
			g.Assert(CleanTemplate("{ANSI:5,0,0}ABC{RGB:0,255,0}CDE{bgHEX:#0000FF}EFG")).Equal("ABCCDEEFG")
		})
	})

	g.Describe("AddCustomStyle / DeleteCustomStyles", func() {
		g.It("Allow to define custom styles, supported both by Colorize and ColorizeTemplate", func() {
			g.Assert(Colorize("ABC", "customRed@@")).Equal("ABC")
			g.Assert(ColorizeTemplate("{customRed@@ green}ABC{-}")).Equal("\x1b[32mABC\x1b[39m\x1b[0m")

			AddCustomStyle("customRed@@", "red", "underline")

			g.Assert(Colorize("ABC", "customRed@@")).Equal("\x1b[31m\x1b[4mABC\x1b[24m\x1b[39m")
			g.Assert(ColorizeTemplate("{customRed@@ green}ABC{-}")).Equal("\x1b[31m\x1b[4m\x1b[32mABC\x1b[24m\x1b[39m\x1b[39m\x1b[0m")

			DeleteCustomStyles("customRed@@")

			g.Assert(Colorize("ABC", "customRed@@")).Equal("ABC")
			g.Assert(ColorizeTemplate("{customRed@@ green}ABC{-}")).Equal("\x1b[32mABC\x1b[39m\x1b[0m")
		})

		g.It("Should reject custom styles name which contain spaces or curly brace", func() {
			for _, s := range []string{"{invalid", "invalid}", "no spaces"} {
				g.Assert(AddCustomStyle(s, "red")).Equal(errors.New("The custom style name should not contain spaces or curly braces"))
			}
		})
	})
}


main.go
/*
 * This file is part of tempera. Copyright (C) 2018 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.
 */

package tempera

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var ansiMatcher = regexp.MustCompile("(?i)^(?:bg)?ansi:(\\d{0,3})(?:[,;](\\d)[,;](\\d))?")
var rgbMatcher = regexp.MustCompile("(?i)^(?:bg)?rgb:(\\d{0,3})[,;](\\d{0,3})[,;](\\d{0,3})")
var hexMatcher = regexp.MustCompile("(?i)^(?:bg)?hex:(?:#?)([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})")

func escapeANSI(codes ...int) string {
	return "\x1b[" + strings.Trim(strings.Replace(fmt.Sprint(codes), " ", ";", -1), "[]") + "m"
}

func specToANSICodes(spec string) ANSICode {
	if strings.HasPrefix(spec, "bg") {
		return ANSIBackgroundCodes
	}

	return ANSIForegroundCodes
}

func parseColorComponent(raw string, base int, min, max int64) (dest int64, err error) {
	dest, err = strconv.ParseInt(raw, base, 0)

	if dest < min || dest > max {
		err = errors.New("Invalid color component")
	}

	return
}

func parseColor(raw string, base int, min, max int64) (r, g, b int64, err error) {
	components := strings.Split(raw, ",")

	if r, err = parseColorComponent(components[0], base, min, max); err != nil {
		return
	}

	if g, err = parseColorComponent(components[1], base, min, max); err != nil {
		return
	}

	if b, err = parseColorComponent(components[2], base, min, max); err != nil {
		return
	}

	return
}

func convertANSIColor(spec ...string) (open, close string) {
	codes := specToANSICodes(spec[0])

	if spec[2] == "" { // Simple color code, range 16 to 255
		color, err := parseColorComponent(spec[1], 10, 16, 255)

		if err == nil {
			open = escapeANSI(codes.Open, 5, int(color))
		}
	} else {
		r, g, b, err := parseColor(strings.Join(spec[1:4], ","), 10, 0, 5)

		if err == nil {
			open = escapeANSI(codes.Open, 5, int(16+36*r+6*g+b))
		}
	}

	close = escapeANSI(codes.Close)
	return
}

func convertRGBColor(base int, spec ...string) (open, close string) {
	codes := specToANSICodes(spec[0])

	// Validate RGB
	r, g, b, err := parseColor(strings.Join(spec[1:4], ","), base, 0, 255)

	if err == nil {
		open = escapeANSI(codes.Open, 2, int(r), int(g), int(b))
		close = escapeANSI(codes.Close)
	}

	return
}

func styleToANSI(style string) (open, close string) {
	if spec := ansiMatcher.FindStringSubmatch(style); len(spec) > 1 { // ANSI 256 colors, either as color number or RGB 0-15 spec
		open, close = convertANSIColor(spec...)
	} else if spec := rgbMatcher.FindStringSubmatch(style); len(spec) > 1 { // ANSI RGB 16m colors
		open, close = convertRGBColor(10, spec...)
	} else if spec := hexMatcher.FindStringSubmatch(style); len(spec) > 1 { // ANSI RGB 16 colors in HEX form
		open, close = convertRGBColor(16, spec...)
	} else if codes, exists := ANSICodes[style]; exists { // Styles tag
		open = escapeANSI(codes.Open)
		close = escapeANSI(codes.Close)
	}

	return
}

// Colorize add colors to a string using a specific list of styles, in order
func Colorize(content string, styles ...string) string {
	var header, footer string

	for _, style := range resolveCustomStyles(styles) {
		open, close := styleToANSI(style)

		if open != "" && close != "" {
			header += open
			footer = close + footer
		}
	}

	return header + content + footer
}


custom.go
/*
 * This file is part of tempera. Copyright (C) 2018 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.
 */

package tempera

import (
	"errors"
	"regexp"
)

var customStyles = make(map[string][]string)
var invalidStyleMatcher = regexp.MustCompile("(?i)[\\s\\{\\}]")

func resolveCustomStyles(styles []string) []string {
	var resolvedStyles = make([]string, 0)

	for _, style := range styles {
		cs, exists := customStyles[style]

		if !exists { // No custom styles, add itself
			resolvedStyles = append(resolvedStyles, style)
		} else {
			resolvedStyles = append(resolvedStyles, cs...)
		}
	}

	return resolvedStyles
}

// AddCustomStyle add a new custom style
func AddCustomStyle(name string, styles ...string) error {
	if invalidStyleMatcher.MatchString(name) {
		return errors.New("The custom style name should not contain spaces or curly braces")
	}

	customStyles[name] = styles

	return nil
}

// DeleteCustomStyles removes one or more custom styles
func DeleteCustomStyles(names ...string) {
	for _, n := range names {
		delete(customStyles, n)
	}
}


template.go
/*
 * This file is part of tempera. Copyright (C) 2018 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.
 */

package tempera

import (
	"regexp"
	"sort"
	"strings"
)

var parser = regexp.MustCompile("(?i)\\{(?:\\{?)[^\\{\\}]+?\\}")
var splitter = regexp.MustCompile("[\\|\\s]+")

func addStyles(replacement *string, tagStyles *[]string, id string) bool {
	open, _ := styleToANSI(id)

	if open == "" { // Validate the style
		return false
	}

	// Push to the the list of the styles of this tag
	*tagStyles = append(*tagStyles, id)

	// Set the style by applying the opening ANSI code
	*replacement += open

	return true
}

func removeStyles(replacement *string, appliedStyles *[][]string) {
	// Pop a the style from the list of the applied ones
	totalStyles := len(*appliedStyles)
	stylesToRemove := (*appliedStyles)[totalStyles-1]
	*appliedStyles = (*appliedStyles)[:totalStyles-1]
	totalStyles--

	for _, id := range stylesToRemove { // Unset the style by applying the closing ANSI codes
		_, close := styleToANSI(id)

		*replacement += close
	}

	if len(*appliedStyles) > 0 { // If the applied styles stack is still non-empty, it means we have to restore the previous style
		for _, id := range (*appliedStyles)[totalStyles-1] {
			open, _ := styleToANSI(id)
			*replacement += open
		}
	}
}

// ColorizeTemplate add colors to a template string
func ColorizeTemplate(template string) string {
	// Create a new styles stack
	appliedStyles := make([][]string, 0)

	// For each tag in the string
	stylesInserted := false
	modified := parser.ReplaceAllStringFunc(template, func(match string) string {
		var replacement string

		if strings.HasPrefix(match, "{{") {
			return match[1:]
		}

		ids := splitter.Split(strings.Trim(match, "{}"), -1)
		tagStyles := make([]string, 0) // Maintain a list of all styles applied in this tag

	TagStyles:
		for _, id := range resolveCustomStyles(ids) { // For each id in this tag
			id = strings.TrimSpace(id)
			switch id {
			case "-": // We are removing a style from the text
				if len(appliedStyles) > 0 { // If there is a previously applied tag
					removeStyles(&replacement, &appliedStyles)
				}

				break TagStyles // Do no process further styles in this tag
			case "reset": // Reset, it means forget all the applied styles so far
				appliedStyles = make([][]string, 0)
				break TagStyles // Do no process further styles in this tag
			default: // We are adding a new style
				if addStyles(&replacement, &tagStyles, id) {
					stylesInserted = true
				}
			}
		}

		// If we added any style in this tag, add to the applied styles stack (in reverse order in order to guarantee proper closing)
		if len(tagStyles) > 0 {
			sort.Sort(sort.Reverse(sort.StringSlice(tagStyles)))
			appliedStyles = append(appliedStyles, tagStyles)
		}

		return replacement
	})

	if stylesInserted { // Always apply the global reset at the end if we apply any style. This is to leave a clean state
		modified += escapeANSI(ANSICodes["reset"].Close)
	}

	return modified
}

// CleanTemplate removes all style tag from a template
func CleanTemplate(template string) string {
	return parser.ReplaceAllStringFunc(template, func(match string) string {
		if strings.HasPrefix(match, "{{") {
			return match[1:]
		}

		return ""
	})
}


codes.go
/*
 * This file is part of tempera. Copyright (C) 2018 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.
 */

package tempera

// ANSICode represents a ANSI code with opening and closing (resetting) code
type ANSICode struct {
	Open  int
	Close int
}

// ANSIForegroundCodes holds general code for foreground color management
var ANSIForegroundCodes = ANSICode{38, 39}

// ANSIBackgroundCodes holds general code for foreground color management
var ANSIBackgroundCodes = ANSICode{48, 49}

// ANSICodes is the list of ANSI Escape codes, taken from https://github.com/chalk/ansi-styles
var ANSICodes = map[string]ANSICode{
	"reset": {0, 0},
	// Style
	"bold":          {1, 22}, // 21 isn't widely supported and 22 does the same thing
	"dim":           {2, 22},
	"italic":        {3, 23},
	"underline":     {4, 24},
	"inverse":       {7, 27},
	"hidden":        {8, 28},
	"strikethrough": {9, 29},
	// Foreground colors
	"black":   {30, 39},
	"red":     {31, 39},
	"green":   {32, 39},
	"yellow":  {33, 39},
	"blue":    {34, 39},
	"magenta": {35, 39},
	"cyan":    {36, 39},
	"white":   {37, 39},
	"gray":    {90, 39},
	// Background colors
	"bgBlack":   {40, 49},
	"bgRed":     {41, 49},
	"bgGreen":   {42, 49},
	"bgYellow":  {43, 49},
	"bgBlue":    {44, 49},
	"bgMagenta": {45, 49},
	"bgCyan":    {46, 49},
	"bgWhite":   {47, 49},
	// Bright foreground colors
	"redBright":     {91, 39},
	"greenBright":   {92, 39},
	"yellowBright":  {93, 39},
	"blueBright":    {94, 39},
	"magentaBright": {95, 39},
	"cyanBright":    {96, 39},
	"whiteBright":   {97, 39},
	// Bright background colors
	"bgBlackBright":   {100, 49},
	"bgRedBright":     {101, 49},
	"bgGreenBright":   {102, 49},
	"bgYellowBright":  {103, 49},
	"bgBlueBright":    {104, 49},
	"bgMagentaBright": {105, 49},
	"bgCyanBright":    {106, 49},
	"bgWhiteBright":   {107, 49},
}
